File: packages/python/onnx9000-core/src/onnx9000/core/models/mae.py
"""Mae."""

from typing import Any

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.models.vit import Block, PatchEmbed
from onnx9000.core.ops import add, concat, flatten, gather, reshape, scatter_nd
from onnx9000.core.primitives import ConvND, Gemm, LayerNormalization


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    """Get param."""
    return Variable(name=name, shape=shape, dtype=dtype)


class MaskedAutoencoderViT:
    """Masked Autoencoder with VisionTransformer backbone."""

    def __init__(  # noqa: D107
        self,
        img_size: int = 224,
        patch_size: int = 16,
        in_chans: int = 3,
        embed_dim: int = 1024,
        depth: int = 24,
        num_heads: int = 16,
        decoder_embed_dim: int = 512,
        decoder_depth: int = 8,
        decoder_num_heads: int = 16,
        mlp_ratio: float = 4.0,
        norm_layer: Any = LayerNormalization,
    ):
        """Init."""
        self.img_size = img_size
        self.patch_size = patch_size
        self.num_patches = (img_size // patch_size) ** 2

        self.embed_dim = embed_dim
        self.decoder_embed_dim = decoder_embed_dim

        # Encoder
        self.patch_embed = PatchEmbed(
            img_size, patch_size, in_chans, embed_dim, prefix="patch_embed"
        )
        self.blocks = [
            Block(embed_dim, num_heads, mlp_ratio, qkv_bias=True, prefix=f"blocks.{i}")
            for i in range(depth)
        ]
        self.norm = norm_layer((embed_dim,))

        # Decoder
        self.decoder_embed = Gemm(trans_b=1)
        self.decoder_blocks = [
            Block(
                decoder_embed_dim,
                decoder_num_heads,
                mlp_ratio,
                qkv_bias=True,
                prefix=f"decoder_blocks.{i}",
            )
            for i in range(decoder_depth)
        ]
        self.decoder_norm = norm_layer((decoder_embed_dim,))
        self.decoder_pred = Gemm(trans_b=1)

    def forward_encoder(self, x: Tensor, mask_indices: Tensor) -> tuple[Tensor, Tensor]:  # noqa: D102
        # Embed patches
        """Forward encoder."""
        x = self.patch_embed(x)

        # Add pos embed w/o cls token
        pos_embed = get_param("pos_embed", [1, self.num_patches + 1, self.embed_dim])

        from onnx9000.core.ops import constant
        from onnx9000.core.ops import slice as slice_op

        starts = constant([1], dtype=7)
        ends = constant([self.num_patches + 1], dtype=7)
        axes = constant([1], dtype=7)
        pos_embed_no_cls = slice_op(pos_embed, starts, ends, axes)

        x = add(x, pos_embed_no_cls)

        # Gather kept patches (we simulate mask by gathering indices)
        x = gather(x, mask_indices, axis=1)

        # Append cls token
        cls_token = get_param("cls_token", [1, 1, self.embed_dim])
        cls_pos_embed = slice_op(
            pos_embed, constant([0], dtype=7), constant([1], dtype=7), constant([1], dtype=7)
        )
        cls_token = add(cls_token, cls_pos_embed)

        x = concat([cls_token, x], axis=1)

        # Apply blocks
        for blk in self.blocks:
            x = blk(x)
        x = self.norm(
            x, get_param("norm.weight", [self.embed_dim]), get_param("norm.bias", [self.embed_dim])
        )

        return x, mask_indices

    def forward_decoder(self, x: Tensor, mask_indices: Tensor) -> Tensor:  # noqa: D102
        # Embed tokens
        """Forward decoder."""
        x = self.decoder_embed(
            x,
            get_param("decoder_embed.weight", [self.decoder_embed_dim, self.embed_dim]),
            get_param("decoder_embed.bias", [self.decoder_embed_dim]),
        )

        # Scatter unmasked tokens back into full sequence
        # ScatterND requires data, indices, updates
        # To reconstruct the full sequence, we start with mask tokens
        mask_token = get_param("mask_token", [1, 1, self.decoder_embed_dim])
        # Expand mask_token to full size
        from onnx9000.core.ops import constant, tile

        full_mask = tile(mask_token, constant([1, self.num_patches, 1], dtype=7))

        # x is [B, 1+kept_len, C], we drop the cls token for scattering
        starts = constant([1], dtype=7)
        ends = constant([self.num_patches + 1], dtype=7)  # just large enough
        axes = constant([1], dtype=7)
        from onnx9000.core.ops import slice as slice_op

        x_no_cls = slice_op(x, starts, ends, axes)

        # Unmask via ScatterND
        x_full = scatter_nd(full_mask, mask_indices, x_no_cls)

        # Add cls token back
        x_cls = slice_op(x, constant([0], dtype=7), constant([1], dtype=7), constant([1], dtype=7))
        x = concat([x_cls, x_full], axis=1)

        # Add pos embed
        decoder_pos_embed = get_param(
            "decoder_pos_embed", [1, self.num_patches + 1, self.decoder_embed_dim]
        )
        x = add(x, decoder_pos_embed)

        # Apply blocks
        for blk in self.decoder_blocks:
            x = blk(x)
        x = self.decoder_norm(
            x,
            get_param("decoder_norm.weight", [self.decoder_embed_dim]),
            get_param("decoder_norm.bias", [self.decoder_embed_dim]),
        )

        # Predict pixels
        x = self.decoder_pred(
            x,
            get_param("decoder_pred.weight", [self.patch_size**2 * 3, self.decoder_embed_dim]),
            get_param("decoder_pred.bias", [self.patch_size**2 * 3]),
        )

        # Drop cls token
        x = slice_op(
            x,
            constant([1], dtype=7),
            constant([self.num_patches + 1], dtype=7),
            constant([1], dtype=7),
        )
        return x

    def __call__(self, x: Tensor, mask_indices: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        latent, mask_indices = self.forward_encoder(x, mask_indices)
        pred = self.forward_decoder(latent, mask_indices)
        return pred


def mae_vit_base_patch16(**kwargs: Any) -> MaskedAutoencoderViT:  # noqa: D103
    """Mae vit base patch16."""
    return MaskedAutoencoderViT(
        patch_size=16,
        embed_dim=768,
        depth=12,
        num_heads=12,
        decoder_embed_dim=512,
        decoder_depth=8,
        decoder_num_heads=16,
        **kwargs,
    )

File: packages/js/core/src/models/mae.ts
import { Tensor } from '../ir/tensor.js';
import { Gemm, LayerNormalization } from '../primitives.js';
import { Block, PatchEmbed } from './vit.js';

function getParam(
  name: string,
  shape: number[],
  dtype: ReturnType<typeof JSON.parse> = 'float32',
): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: ReturnType<typeof JSON.parse>): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class MaskedAutoencoderViT {
  public imgSize: number;
  public patchSize: number;
  public numPatches: number;
  public embedDim: number;
  public decoderEmbedDim: number;

  public patchEmbed: PatchEmbed;
  public blocks: Block[];
  public norm: LayerNormalization;

  public decoderEmbed: Gemm;
  public decoderBlocks: Block[];
  public decoderNorm: LayerNormalization;
  public decoderPred: Gemm;

  constructor(
    imgSize: number = 224,
    patchSize: number = 16,
    inChans: number = 3,
    embedDim: number = 1024,
    depth: number = 24,
    numHeads: number = 16,
    decoderEmbedDim: number = 512,
    decoderDepth: number = 8,
    decoderNumHeads: number = 16,
    mlpRatio: number = 4.0,
  ) {
    this.imgSize = imgSize;
    this.patchSize = patchSize;
    this.numPatches = Math.floor(imgSize / patchSize) * Math.floor(imgSize / patchSize);

    this.embedDim = embedDim;
    this.decoderEmbedDim = decoderEmbedDim;

    this.patchEmbed = new PatchEmbed(imgSize, patchSize, inChans, embedDim, 'patch_embed');
    this.blocks = [];
    for (let i = 0; i < depth; i++) {
      this.blocks.push(new Block(embedDim, numHeads, mlpRatio, true, `blocks.${i}`));
    }
    this.norm = new LayerNormalization([embedDim]);

    this.decoderEmbed = new Gemm(1.0, 1.0, 0, 1);
    this.decoderBlocks = [];
    for (let i = 0; i < decoderDepth; i++) {
      this.decoderBlocks.push(
        new Block(decoderEmbedDim, decoderNumHeads, mlpRatio, true, `decoder_blocks.${i}`),
      );
    }
    this.decoderNorm = new LayerNormalization([decoderEmbedDim]);
    this.decoderPred = new Gemm(1.0, 1.0, 0, 1);
  }

  forwardEncoder(x: Tensor, maskIndices: Tensor): [Tensor, Tensor] {
    x = this.patchEmbed.call(x);

    const posEmbed = getParam('pos_embed', [1, this.numPatches + 1, this.embedDim]);

    const starts = recordOp('Constant', [], { value: [1], dtype: 7 });
    const ends = recordOp('Constant', [], { value: [this.numPatches + 1], dtype: 7 });
    const axes = recordOp('Constant', [], { value: [1], dtype: 7 });
    const posEmbedNoCls = recordOp('Slice', [posEmbed, starts, ends, axes]);

    x = recordOp('Add', [x, posEmbedNoCls]);

    x = recordOp('Gather', [x, maskIndices], { axis: 1 });

    let clsToken = getParam('cls_token', [1, 1, this.embedDim]);
    const clsPosEmbed = recordOp('Slice', [
      posEmbed,
      recordOp('Constant', [], { value: [0], dtype: 7 }),
      recordOp('Constant', [], { value: [1], dtype: 7 }),
      recordOp('Constant', [], { value: [1], dtype: 7 }),
    ]);
    clsToken = recordOp('Add', [clsToken, clsPosEmbed]);

    x = recordOp('Concat', [clsToken, x], { axis: 1 });

    for (const blk of this.blocks) {
      x = blk.call(x);
    }
    x = this.norm.call(
      x,
      getParam('norm.weight', [this.embedDim]),
      getParam('norm.bias', [this.embedDim]),
    );

    return [x, maskIndices];
  }

  forwardDecoder(x: Tensor, maskIndices: Tensor): Tensor {
    x = this.decoderEmbed.call(
      x,
      getParam('decoder_embed.weight', [this.decoderEmbedDim, this.embedDim]),
      getParam('decoder_embed.bias', [this.decoderEmbedDim]),
    );

    const maskToken = getParam('mask_token', [1, 1, this.decoderEmbedDim]);
    const fullMask = recordOp('Tile', [
      maskToken,
      recordOp('Constant', [], { value: [1, this.numPatches, 1], dtype: 7 }),
    ]);

    const starts = recordOp('Constant', [], { value: [1], dtype: 7 });
    const ends = recordOp('Constant', [], { value: [this.numPatches + 1], dtype: 7 });
    const axes = recordOp('Constant', [], { value: [1], dtype: 7 });
    const xNoCls = recordOp('Slice', [x, starts, ends, axes]);

    const xFull = recordOp('ScatterND', [fullMask, maskIndices, xNoCls]);

    const xCls = recordOp('Slice', [
      x,
      recordOp('Constant', [], { value: [0], dtype: 7 }),
      recordOp('Constant', [], { value: [1], dtype: 7 }),
      recordOp('Constant', [], { value: [1], dtype: 7 }),
    ]);
    x = recordOp('Concat', [xCls, xFull], { axis: 1 });

    const decoderPosEmbed = getParam('decoder_pos_embed', [
      1,
      this.numPatches + 1,
      this.decoderEmbedDim,
    ]);
    x = recordOp('Add', [x, decoderPosEmbed]);

    for (const blk of this.decoderBlocks) {
      x = blk.call(x);
    }
    x = this.decoderNorm.call(
      x,
      getParam('decoder_norm.weight', [this.decoderEmbedDim]),
      getParam('decoder_norm.bias', [this.decoderEmbedDim]),
    );

    x = this.decoderPred.call(
      x,
      getParam('decoder_pred.weight', [this.patchSize * this.patchSize * 3, this.decoderEmbedDim]),
      getParam('decoder_pred.bias', [this.patchSize * this.patchSize * 3]),
    );

    x = recordOp('Slice', [
      x,
      recordOp('Constant', [], { value: [1], dtype: 7 }),
      recordOp('Constant', [], { value: [this.numPatches + 1], dtype: 7 }),
      recordOp('Constant', [], { value: [1], dtype: 7 }),
    ]);
    return x;
  }

  call(x: Tensor, maskIndices: Tensor): Tensor {
    let latent;
    [latent, maskIndices] = this.forwardEncoder(x, maskIndices);
    const pred = this.forwardDecoder(latent, maskIndices);
    return pred;
  }
}

export function maeVitBasePatch16(): MaskedAutoencoderViT {
  return new MaskedAutoencoderViT(224, 16, 3, 768, 12, 12, 512, 8, 16, 4.0);
}
